core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -199,6 +198,7 @@ private[kafka] object LogValidator extends Logging {
 
     for (batch <- records.batches.asScala) {
       if (batch.magic() >= RecordBatch.MAGIC_VALUE_V2) {
+        // note since we only check for v2 and beyond, we should always assume one batch records.
         validateOneBatchRecords(records)
```

Comment:
It's not too big of a deal, but we're a little inconsistent on the use of `validateOneBatchRecords`. Here we use it inside the loop over the batches, but below we do it before the loop.

Reply:
The reason is that we cannot get the magic byte until we get to the batch. As you mentioned in the other comment we can for example just check the first batch's magic number. Will refactor a bit.